examples/other/image_histogram.py

```diff
@@ -0,0 +1,582 @@
+# different workgroups don't get in each-others way. After that. each
+# thread in a workgroup is then responsible for adding a single
+# histogram value to the final histogram buffer.
+#
+# The second shader simply copies values from the histogram into the
+# positions buffer, where we can use 1024 cores, so it's parellized
+# pretty well.
+#
+# Took inspiration from https://webgpufundamentals.org/webgpu/lessons/webgpu-compute-shaders-histogram.html
+# but implemented a slightly simpler variant of its final one.
+# It is very likely that a more performant implementation exists. But it's also
+# likely that what its the most efficient version depends on the hardware.
+#
+# About workgroup/invocation id's:
+#
+# * local_invocation_id: the vec3 indicating the current invocation into the
+#   workgroup, as specified using @workgroup_size,  i.e. its position in the
+#   workgroup grid.
+# * local_invocation_index: the u32 represening the 'flat' local_invocation_id.
+# * workgroup_id: the vec3 indicating the position of the workgroup in overall
+#   compute shader grid, as specified by dispatch_workgroups().
+# * global_invocation_id: workgroup_id * workgroup_size + local_invocation_id.
+
+
+histogram_wgsl = """
+
+const chunkWidth = 16u;
+const chunkHeight = 16u;
+const chunkDepth = 4u;
+
+const chunkSize = chunkWidth * chunkHeight * chunkDepth;
+const binCount = chunkWidth * chunkHeight;
+var<workgroup> bins: array<atomic<u32>, chunkSize>;
+
+
+@group(0) @binding(0) var imageTexture: texture_2d<f32>;
+@group(0) @binding(1) var<storage, read_write> sa_bins: array<atomic<u32>>;
+
+const kSRGBLuminanceFactors = vec3f(0.2126, 0.7152, 0.0722);
+fn srgbLuminance(color: vec3f) -> f32 {
+    return saturate(dot(color, kSRGBLuminanceFactors));
+}
+
+
+@compute @workgroup_size(chunkWidth, chunkHeight, chunkDepth)
+fn calc_histogram(
+    @builtin(global_invocation_id) global_invocation_id: vec3u,
+    @builtin(local_invocation_index) local_invocation_index: u32,
+    @builtin(local_invocation_id) local_invocation_id: vec3u,
+) {
+
+    // Write zeros to the workgroup array
+    let binIndex = local_invocation_index;
+    atomicStore(&bins[binIndex], 0u);
+
+    workgroupBarrier();
+
+    // Collect pixels, and store in workgroup array
+    let size = textureDimensions(imageTexture, 0);
+    let position = global_invocation_id.xy;
+    if (all(position < size)) {
+        let lastBinIndex = u32(binCount - 1);
+        let color = textureLoad(imageTexture, position, 0);
+        let channel = u32(local_invocation_id.z);
+        var v: f32;
+        if (channel == 3u) {
+            v = srgbLuminance(color.rgb);
+        } else {
+            v = color[channel];
+        }
+        let binIndex = min(u32(v * f32(binCount)), lastBinIndex) * 4u + channel;
+        atomicAdd(&bins[binIndex], 1u);
+    }
+
+    workgroupBarrier();
+
+    // Copy the bin values from our workgroup array to the storage buffer.
+    let binValue = atomicLoad(&bins[binIndex]);
```

Comment:
Why do you need an atomic load here? Doesn't this thread own the variable?

Reply:
The example at webgpufundamentals does it too. I'll have another look. I think I also changed back and forth a bit as I was trying to get things to work.

Reply:
I think this is because `bins` is a workgroup variable shared among all threads within the workgroup. Therefore, in a multithreaded context, its read and write operations need to be atomic to ensure data consistency.

Reply:
In this (and one other) case there is no risk on race-conditions, because each thread only touches its own index. I can indeed use normal array opeations, and that works. However, the WGSL spec states:

> the only valid operations on atomic objects are the [atomic builtin functions](https://www.w3.org/TR/WGSL/#atomic-builtin-functions).

So, I expect that at some point that will be forbidden. Note that atomicity is address specific, to the fact that we have `atomicLoad` here does not mean that the different threads wait for each-other, since they never touch the same address.